File: backend/app/calendar.py
"""
Calendar provider abstraction.

A single Protocol with Google and Microsoft adapters. The adapters are
SCAFFOLDING: the request-shaping and boundary are real, but the actual HTTP
calls are gated behind `is_configured()` and raise CalendarNotConfigured until
OAuth credentials exist in the environment. This lets the async sync pipeline
(outbox -> worker -> provider -> meeting_calendar_links) be built and exercised
end-to-end now, and turned live later by only filling in `_create_event` and the
env vars — no call-site changes.

Provider tokens live only in the backend env; they are never returned to clients.
"""
from typing import Protocol

from app import config


class CalendarNotConfigured(Exception):
    """Raised when a provider is selected but its OAuth credentials are absent."""


class CalendarEventResult(dict):
    """{'external_event_id':..., 'external_calendar_id':...}"""


class CalendarProvider(Protocol):
    name: str
    async def create_event(self, *, title: str, description: str | None,
                           start, end, attendees: list[str]) -> CalendarEventResult: ...
    async def update_event(self, external_event_id: str, **fields) -> CalendarEventResult: ...
    async def cancel_event(self, external_event_id: str) -> None: ...


class _BaseProvider:
    name = "base"
    def is_configured(self) -> bool:                       # pragma: no cover - trivial
        return False
    async def create_event(self, *, title, description, start, end, attendees):
        if not self.is_configured():
            raise CalendarNotConfigured(f"{self.name} calendar is not configured")
        return await self._create_event(title, description, start, end, attendees)
    async def update_event(self, external_event_id, **fields):
        if not self.is_configured():
            raise CalendarNotConfigured(f"{self.name} calendar is not configured")
        return await self._update_event(external_event_id, **fields)
    async def cancel_event(self, external_event_id):
        if not self.is_configured():
            raise CalendarNotConfigured(f"{self.name} calendar is not configured")
        await self._cancel_event(external_event_id)
    # ---- to implement when going live (real HTTP + OAuth token refresh) ----
    async def _create_event(self, title, description, start, end, attendees):  # pragma: no cover
        raise NotImplementedError
    async def _update_event(self, external_event_id, **fields):                # pragma: no cover
        raise NotImplementedError
    async def _cancel_event(self, external_event_id):                          # pragma: no cover
        raise NotImplementedError


class GoogleCalendarProvider(_BaseProvider):
    name = "google"
    def is_configured(self) -> bool:
        return bool(getattr(config, "GOOGLE_CALENDAR_CLIENT_ID", None)
                    and getattr(config, "GOOGLE_CALENDAR_REFRESH_TOKEN", None))
    # _create_event etc. would POST to https://www.googleapis.com/calendar/v3/... here.


class MicrosoftOutlookProvider(_BaseProvider):
    name = "microsoft"
    def is_configured(self) -> bool:
        return bool(getattr(config, "MS_CALENDAR_CLIENT_ID", None)
                    and getattr(config, "MS_CALENDAR_REFRESH_TOKEN", None))
    # _create_event etc. would POST to https://graph.microsoft.com/v1.0/me/events here.


_PROVIDERS = {p.name: p for p in (GoogleCalendarProvider(), MicrosoftOutlookProvider())}


def get_provider(name: str | None) -> _BaseProvider | None:
    return _PROVIDERS.get((name or "").lower())

File: apps/backend-node/src/integrations/calendar.ts
/**
 * Calendar provider abstraction. Faithful port of backend/app/calendar.py.
 *
 * Google and Microsoft adapters are SCAFFOLDING: the request-shaping and
 * boundary are real, but the actual HTTP calls are gated behind isConfigured()
 * and throw CalendarNotConfigured until OAuth creds exist. This lets the async
 * sync pipeline (outbox → worker → provider → meeting_calendar_links) be
 * exercised end-to-end now and turned live later by only filling in _createEvent
 * and the env vars. Reads *_CALENDAR_* from process.env at call time so it works
 * in both the Nest app and the worker. Provider tokens never reach clients.
 */
export class CalendarNotConfigured extends Error {}

export interface CalendarEventResult {
  external_event_id?: string;
  external_calendar_id?: string;
}

export interface CreateEventArgs {
  title?: string | null;
  description?: string | null;
  start?: string | null;
  end?: string | null;
  attendees?: string[];
}

abstract class BaseProvider {
  abstract readonly name: string;
  isConfigured(): boolean {
    return false;
  }
  async createEvent(args: CreateEventArgs): Promise<CalendarEventResult> {
    if (!this.isConfigured()) throw new CalendarNotConfigured(`${this.name} calendar is not configured`);
    return this._createEvent(args);
  }
  async updateEvent(externalEventId: string, fields: Record<string, unknown>): Promise<CalendarEventResult> {
    if (!this.isConfigured()) throw new CalendarNotConfigured(`${this.name} calendar is not configured`);
    return this._updateEvent(externalEventId, fields);
  }
  async cancelEvent(externalEventId: string): Promise<void> {
    if (!this.isConfigured()) throw new CalendarNotConfigured(`${this.name} calendar is not configured`);
    await this._cancelEvent(externalEventId);
  }
  // ---- to implement when going live (real HTTP + OAuth token refresh) ----
  protected async _createEvent(_args: CreateEventArgs): Promise<CalendarEventResult> {
    throw new Error('not implemented');
  }
  protected async _updateEvent(_id: string, _fields: Record<string, unknown>): Promise<CalendarEventResult> {
    throw new Error('not implemented');
  }
  protected async _cancelEvent(_id: string): Promise<void> {
    throw new Error('not implemented');
  }
}

class GoogleCalendarProvider extends BaseProvider {
  readonly name = 'google';
  isConfigured(): boolean {
    return Boolean(process.env.GOOGLE_CALENDAR_CLIENT_ID && process.env.GOOGLE_CALENDAR_REFRESH_TOKEN);
  }
  // _createEvent would POST to https://www.googleapis.com/calendar/v3/... here.
}

class MicrosoftOutlookProvider extends BaseProvider {
  readonly name = 'microsoft';
  isConfigured(): boolean {
    return Boolean(process.env.MS_CALENDAR_CLIENT_ID && process.env.MS_CALENDAR_REFRESH_TOKEN);
  }
  // _createEvent would POST to https://graph.microsoft.com/v1.0/me/events here.
}

const PROVIDERS: Record<string, BaseProvider> = {
  google: new GoogleCalendarProvider(),
  microsoft: new MicrosoftOutlookProvider(),
};

export function getProvider(name: string | null | undefined): BaseProvider | null {
  return PROVIDERS[(name || '').toLowerCase()] ?? null;
}
